### PR TITLE
Homogenization of the providers (part 1)

### PR DESCRIPTION
**Pull request by ldez**
### What does this PR do?

The first step a the homogenization of the providers.

#### Rewrite Labels/Annotations system (provider with template)

- [x] Docker: `map[string]string`
- [x] Eureka: `map[string]string`
- [x] k8s: `map[string]string`
- [x] Marathon: `*map[string]string`
- [x] Rancher: `map[string]string`
- [x] Mesos: `task.Labels`
- [x] ecs: `map[string]*string`

more:
- [x] use label package (function and constant)
- [x] remove dead tests
- [x] move the build of funcMap on a dedicate file
- [x] add tests on label package

#### No-Rewrite

- KV: BoldDB, Consul, Etcd, ZK
- DynamoDB (no template)

### Motivation

make it easier to add new label.

### More

- [x] Added/updated tests
- [x] Added/updated documentation

Mjolnir

### CODE REVIEW

File: provider/marathon/config.go [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+	return method || sticky || stickiness
+}
+
+func hasMaxConnLabels(application marathon.Application) bool {
+	mca := label.HasP(application.Labels, label.TraefikBackendMaxconnAmount)
+	mcef := label.HasP(application.Labels, label.TraefikBackendMaxconnExtractorfunc)
+	return mca && mcef
+}
+
+func getSticky(application marathon.Application) string {
```
Comment (mmatur) [inline provider/marathon/config.go]: Could you please add a comment 
```go 
// TODO: Deprecated
// Deprecated replaced by Stickiness
```

File: provider/rancher/config.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+func (p *Provider) getFrontendRule(service rancherData) string {
+	defaultRule := "Host:" + strings.ToLower(strings.Replace(service.Name, "/", ".", -1)) + "." + p.Domain
+	return label.GetStringValue(service.Labels, label.TraefikFrontendRule, defaultRule)
+}
+
+func (p *Provider) getFrontendName(service rancherData) string {
+	return provider.Normalize(p.getFrontendRule(service))
+}
+
+func getSticky(service rancherData) string {
```
Comment (mmatur) [inline provider/rancher/config.go]: Could you please add a comment 
```go 
// TODO: Deprecated
// Deprecated replaced by Stickiness
```

File: provider/ecs/config.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+		services,
+	})
+}
+
+func (p *Provider) getFrontendRule(i ecsInstance) string {
+	defaultRule := "Host:" + strings.ToLower(strings.Replace(i.Name, "_", "-", -1)) + "." + p.Domain
+	return getStringValue(i, label.TraefikFrontendRule, defaultRule)
+}
+
+func getSticky(instances []ecsInstance) string {
```
Comment (mmatur) [inline provider/ecs/config.go]: Could you please add a comment 
```go 
// TODO: Deprecated
// Deprecated replaced by Stickiness
```

File: provider/docker/config_container.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+	return ""
+}
+
+// Escape beginning slash "/", convert all others to dash "-", and convert underscores "_" to dash "-"
+func getSubDomain(name string) string {
+	return strings.Replace(strings.Replace(strings.TrimPrefix(name, "/"), "/", "-", -1), "_", "-", -1)
+}
+
+// Deprecated replaced by Stickiness
+func getSticky(container dockerData) string {
```
Comment (mmatur) [inline provider/docker/config_container.go]: Could you please add a comment 
```go 
// TODO: Deprecated
// Deprecated replaced by Stickiness
```

File: provider/marathon/builder_test.go [hunk 5]
```diff
@@ -55,9 +55,9 @@ func labelWithService(key, value string, serviceName string) func(*marathon.Appl
 		panic("serviceName can not be empty")
 	}
 
-	property := strings.TrimPrefix(key, types.LabelPrefix)
+	property := strings.TrimPrefix(key, label.Prefix)
 	return func(app *marathon.Application) {
-		app.AddLabel(types.LabelPrefix+serviceName+"."+property, value)
+		app.AddLabel(label.Prefix+serviceName+"."+property, value)
```
Comment (nmengin) [inline provider/marathon/builder_test.go]: Should not we use `fmt.Sprintf`?

File: provider/consul/consul_catalog_config.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+func (p *CatalogProvider) getFrontendRule(service serviceUpdate) string {
+	customFrontendRule := p.getAttribute("frontend.rule", service.Attributes, "")
+	if customFrontendRule == "" {
+		customFrontendRule = p.FrontEndRule
+	}
+
+	tmpl := p.frontEndRuleTemplate
+	tmpl, err := tmpl.Parse(customFrontendRule)
+	if err != nil {
+		log.Errorf("failed to parse Consul Catalog custom frontend rule: %s", err)
```
Comment (nmengin) [inline provider/consul/consul_catalog_config.go]: Can you upper case it?

File: provider/consul/consul_catalog_config.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	}{
+		ServiceName: service.ServiceName,
+		Domain:      p.Domain,
+		Attributes:  service.Attributes,
+	}
+
+	var buffer bytes.Buffer
+	err = tmpl.Execute(&buffer, templateObjects)
+	if err != nil {
+		log.Errorf("failed to execute Consul Catalog custom frontend rule template: %s", err)
```
Comment (nmengin) [inline provider/consul/consul_catalog_config.go]: Can you upper case it?

File: provider/marathon/config.go [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+		Applications []marathon.Application
+		Domain       string // FIXME DEAD ?
+	}{
+		Applications: filteredApps,
+		Domain:       p.Domain, // FIXME DEAD ?
+	}
+
+	configuration, err := p.GetConfiguration("templates/marathon.tmpl", MarathonFuncMap, templateObjects)
+	if err != nil {
+		log.Errorf("failed to render Marathon configuration template: %s", err)
```
Comment (nmengin) [inline provider/marathon/config.go]: Can you upper case it?

File: provider/marathon/config.go [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+	}
+
+	if p.MarathonLBCompatibility {
+		if value := label.GetStringValueP(application.Labels, labelLbCompatibility, ""); len(value) > 0 {
+			return "Host:" + value
+		}
+	}
+
+	if len(serviceName) > 0 {
+		return "Host:" + strings.ToLower(provider.Normalize(serviceName)) + "." + p.getSubDomain(application.ID) + "." + p.Domain
```
Comment (nmengin) [inline provider/marathon/config.go]: Should not we use `fmt.Sprintf`?

File: provider/marathon/config.go [hunk 10]
```diff
@@ -0,0 +1,416 @@
...
+	}
+	return "Host:" + p.getSubDomain(application.ID) + "." + p.Domain
```
Comment (nmengin) [inline provider/marathon/config.go]: Should not we use `fmt.Sprintf`?

File: provider/marathon/config.go [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+}
+
+func (p *Provider) getBackend(application marathon.Application, serviceName string) string {
+	labels := getLabels(application, serviceName)
+	lblBackend := getLabelName(serviceName, label.SuffixBackend)
+	value := label.GetStringValue(labels, lblBackend, "")
+	if len(value) > 0 {
+		return "backend" + value
+	}
+	return provider.Normalize("backend" + application.ID + getServiceNameSuffix(serviceName))
```
Comment (nmengin) [inline provider/marathon/config.go]: Should not we use `fmt.Sprintf`?

File: provider/marathon/config.go [hunk 12]
```diff
@@ -0,0 +1,416 @@
...
+}
+
+func (p *Provider) getFrontendName(application marathon.Application, serviceName string) string {
+	return provider.Normalize("frontend" + application.ID + getServiceNameSuffix(serviceName))
```
Comment (nmengin) [inline provider/marathon/config.go]: Should not we use `fmt.Sprintf`?

File: provider/mesos/config.go [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+	return ""
+}
+
+// getFrontendRule returns the frontend rule for the specified application, using
+// it's label. It returns a default one (Host) if the label is not present.
+func (p *Provider) getFrontendRule(task state.Task) string {
+	if v := getStringValue(task, label.TraefikFrontendRule, ""); len(v) > 0 {
+		return v
+	}
+	return "Host:" + strings.ToLower(strings.Replace(p.getSubDomain(task.DiscoveryInfo.Name), "_", "-", -1)) + "." + p.Domain
```
Comment (nmengin) [inline provider/mesos/config.go]: Should not we use `fmt.Sprintf`?

File: provider/mesos/config_test.go [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+		}
+		actualConfig := provider.buildConfiguration()
+		if c.expectedNil {
+			if actualConfig != nil {
+				t.Fatalf("Should have been nil, got %v", actualConfig)
+			}
+		} else {
+			// Compare backends
+			if !reflect.DeepEqual(actualConfig.Backends, c.expectedBackends) {
+				t.Fatalf("expected %#v, got %#v", c.expectedBackends, actualConfig.Backends)
```
Comment (nmengin) [inline provider/mesos/config_test.go]: Can you upper case it?

File: provider/mesos/config_test.go [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+			}
+			if !reflect.DeepEqual(actualConfig.Frontends, c.expectedFrontends) {
+				t.Fatalf("expected %#v, got %#v", c.expectedFrontends, actualConfig.Frontends)
```
Comment (nmengin) [inline provider/mesos/config_test.go]: Can you upper case it?

File: provider/mesos/config_test.go [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+		Hostname: "127.0.0.1",
+	}
+	var taskState = state.State{
+		Slaves:     []state.Slave{slave},
+		Frameworks: []state.Framework{framework},
+	}
+
+	var p = taskRecords(taskState)
+	if len(p) == 0 {
+		t.Fatal("taskRecord should return at least one task")
```
Comment (nmengin) [inline provider/mesos/config_test.go]: Can you add the word `Function` at the beginning,? Thus the falal log begins with an upper case.

File: provider/rancher/config.go [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+			log.Debugf("Filtering service %s with state of %s", service.Name, service.State)
+			return false
+		}
+	}
+
+	return true
+}
+
+func (p *Provider) getFrontendRule(service rancherData) string {
+	defaultRule := "Host:" + strings.ToLower(strings.Replace(service.Name, "/", ".", -1)) + "." + p.Domain
```
Comment (nmengin) [inline provider/rancher/config.go]: Should not we use `fmt.Sprintf`?

File: provider/consul/consul_catalog.go [hunk 18]
```diff
@@ -59,63 +65,149 @@ func (a nodeSorter) Swap(i int, j int) {
...
+		for {
+			select {
+			case <-stopCh:
+				return
+			default:
+			}
+
+			data, meta, err := catalog.Services(options)
+			if err != nil {
+				log.Errorf("Failed to list services: %s", err)
```
Comment (nmengin) [inline provider/consul/consul_catalog.go]: Should not we use `%v` instead of `%s` to log an `err`?

File: provider/consul/consul_catalog.go [hunk 19]
```diff
@@ -59,63 +65,149 @@ func (a nodeSorter) Swap(i int, j int) {
...
+			}
+
+			options.WaitIndex = meta.LastIndex
+
+			if data != nil {
+				current := make(map[string]Service)
+				for key, value := range data {
+					nodes, _, err := catalog.Service(key, "", &api.QueryOptions{})
+					if err != nil {
+						log.Errorf("Failed to get detail of service %s: %s", key, err)
```
Comment (nmengin) [inline provider/consul/consul_catalog.go]: Should not we use `%v` instead of `%s` to log an `err`?

File: provider/marathon/config.go (hunk 9 above)
Comment (ldez) [reply provider/marathon/config.go]: because it's not Java  :stuck_out_tongue_closed_eyes:

Comment (nmengin) [reply provider/marathon/config.go]: But, IMHO, it's easier to read. :smile_cat:

**Comment by emilevauge** [comment]
:cry: I failed to review this PR, but that's awesome job @ldez :clap: :clap: :clap:

**Pull request merged**